### ISSUE TITLE
bindings/python: add Workflow class

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#2653]
Could we provide a high level `Workflow` class that provides a non-reactive interface to submitting and waiting for jobs all together?  When you instantiate it, you could tell it what states you care about, and it could subscribe to `job-state` events.  For each job you submit via the class, it would internally cache a future.  After submitting some jobs, you could call a blocking "wait" method that would enter the reactor and process both submit responses and `job-state` events as they occur.  

Wait could return a (jobid, state) tuple or throw an exception e.g. if submit fails.  It could return `None` when no more notifications are possible for the requested states.  You could alternately call submit(s) and wait(s) to implement a workflow logic without exposing the complexity of the flux reactor.

Maybe it could fetch a job's eventlog on completion and then tack on a result to the tuple for the INACTIVE state transition (jobid, state, success, reason).

Maybe it could use the lightweight `job-manager.wait` RPC if you tell it up front that you only care about  DEPEND (implied by submit response) and INACTIVE states.

Maybe it could use the heavyweight, per-job eventlog watch if you tell it you want even more detail about what's going on (or perhaps if an alternate waitevent method is called).

_Originally posted by @garlick in https://github.com/flux-framework/flux-core/issues/2652#issuecomment-575700038_

**Comment by SteVwonder** [comment]
I like this idea!  I think we could also considering adding a `Job` class that has many of the same interfaces you mentioned but for just a single job, and then a `Workflow` class could be composed of/contain many `Job`s.

>  After submitting some jobs, you could call a blocking "wait" method that would enter the reactor and process both submit responses and job-state events as they occur.

The responses that come back are queueing up at the client, right?  So if a `Workflow` user delays calling the `wait` method for an extended period of time, that will only hurt their individual/local connector's performance and not the whole instance, right?

> Wait could return a (jobid, state) tuple or throw an exception e.g. if submit fails. 

The only thing that I am hung up on about throwing a python exception for `Workflow.wait()` is that many jobs within the workflow could have failed for various reasons, but you can only throw a single exception.  So I am leaning towards returning the tuple or storing the results inside the `Workflow` class and providing an iterface for the user to pull that info out/query it.

> Maybe it could use the lightweight job-manager.wait RPC if you tell it up front that you only care about DEPEND (implied by submit response) and INACTIVE states.
> Maybe it could use the heavyweight, per-job eventlog watch if you tell it you want even more detail about what's going on (or perhaps if an alternate waitevent method is called).

I really like the idea of abstracting this away from the user.  I need to noodle some more on the best way to have the user tell us we need to track more than just DEPEND and INACTIVE.  It could just simply be an optional argument to the `Workflow` constructor and/or a method on the `Workflow` class that switches from lightweight to heavyweight.

I'm not 100% sold on the strawman `followed_by` interface that I've used below, but for the sake of conversation, I think an interface along these lines were be pretty cool.  Thoughts?

```python
setup_job = JobspecV1.from_command('foo')
compute_jobs = [JobspecV1.from_command(x) for x in ['qux', 'quux', 'quuz']]
postprocess_job = JobspecV1.from_command('bar')
workflow = setup_job.followed_by(*compute_jobs).followed_by(postprocess_job)

workflow.submit()
# do stuff here; could check for submission failures, poke at jobs within workflow, or do something else entirely
workflow.wait()
for job in workflow.jobs(): # could also have workflow.failures() and workflow.successes()
  if job.success == False:
    print("Job {} failed with reason: {}".format(job.jobid, job.reason))
    print("\t Eventlog: {}".format(job.eventlog))
  else:
    print("Job {} completed successfully".format(job.jobid))
```

**Comment by garlick** [comment]
> The responses that come back are queueing up at the client, right? So if a Workflow user delays calling the wait method for an extended period of time, that will only hurt their individual/local connector's performance and not the whole instance, right?

Yes, and then only if they are doing synchronous RPCs.  They could do some non-flux stuff for example, or issue non-synchronous workflow.submits.

I like where you're going with your example, although your wait looks like a "wait all"?   That would be useful to have too.

I was picturing wait would return once per state transition, per job.  If there was a job object (good idea IMHO!), then I would imagine wait returns one of those objects, and then you could call various non-blocking job accessors in the body of the loop (get state, get events, get stdout, etc).  The workflow object would have asynchronously fetched what you're interested in before returning from wait (since you told it to earlier), and if you call an accessor inappropriate for the state, then you'd get `None` or similar, not synchronous behavior.

**Comment by SteVwonder** [comment]
> I was picturing wait would return once per state transition, per job. If there was a job object (good idea IMHO!), then I would imagine wait returns one of those objects, and then you could call various non-blocking job accessors in the body of the loop (get state, get events, get stdout, etc). The workflow object would have asynchronously fetched what you're interested in before returning from wait (since you told it to earlier), and if you call an accessor inappropriate for the state, then you'd get None or similar, not synchronous behavior.

Ah! That makes a lot of sense.  I think I'm on the same page as you now.  It sounds like the types of semantics we want are:

- Job class:
  - wait for the job to complete
  - wait for the next job event
  - wait for the next job state transition
    - This may be the same as the previous item, or we may decide we don't want to expose states at all.  I still don't think I fully grok our "job state story"
- Workflow class:
  - wait for all jobs in the workflow to complete
  - wait until all jobs in the workflow to have a particular event in their eventlog
  - wait for the next event for any job in the workflow
  - wait for the next job state transition for any job in the workflow
    - same state/event caveat as above

Does that seem right?

**Comment by grondo** [comment]
I don't want to throw a wrench in the discussion here, but I had two thoughts that are likely just dumb, but I thought in my loneliness I'd share them anyway:

 1. I feel like Workflow is an overloaded term and implies more functionality than waitany/waitall. (e.g. dependencies, data flow. The Workflow name does make some sense here, but I'm wondering if we could come up with a more specific term that would also be acceptable. It seems like what we're talking about is a JobPool or Ensemble.

2. It really feels like Job could inherit from Workflow or vice versa. Couldn't a Workflow itself contain another Workflow, perhaps represented by a sub-instance, which itself is just a Job? It would be neat if the recursion of Flux was somehow represented in this interface, though perhaps that is for something a bit higher level. (I told you these were dumb ideas)

**Comment by SteVwonder** [comment]
> I feel like Workflow is an overloaded term and implies more functionality than waitany/waitall.  (e.g. dependencies, data flow, ...).

Yeah, that's totally fair. We don't want the interface to be misconstrued as a full on workflow manager. I think we could pretty easily include dependencies in the design (once the `depend` module exists).  We will probably care more about data flow with the emergence of more tiered storage systems but not at the level that most workflow managers care.

> The Workflow name does make some sense here, but I'm wondering if we could come up with a more specific term that would also be acceptable. It seems like what we're talking about is a JobPool or Ensemble.

  I like the `Ensemble` and `JobPool` suggestions.  Looping back to the idea of dependencies, what do you think of `DAG` or `JobDAG`?

> It really feels like Job could inherit from Workflow or vice versa. Couldn't a Workflow itself contain another Workflow, perhaps represented by a sub-instance, which itself is just a Job?

Yeah, I agree.  With the strawman `followed_by` method above, I was thinking it would be present as a method in both `Job` and `Workflow`  (or whatever name we decide), accept as an argument an array of either `Job` `Workflow`, and always return a `Workflow`.  

>  It would be neat if the recursion of Flux was somehow represented in this interface, though perhaps that is for something a bit higher level. (I told you these were dumb ideas)

I hadn't thought about nested instances.  That's a great idea.

**Comment by grondo** [comment]
If dependencies are first class part of the interface, then I actually don't mind `Workflow` at all. I had misconstrued that we were only discussing an interface that can `wait` on jobs in bulk.
